### ISSUE TITLE
Project future

### ISSUE THREAD

**Issue by goanpeca** [lektor/lektor#724]
Hello fellow @lektor/developers 

I have not been able to help the project in any meaningful way so far (just lack of time to be honest 😭) But I am a heavy user of it!

I would like to propose something that could help the project move forward in a number of ways. I suggest we create a Lektor opencollective. On the [spyder project](https://opencollective.com/spyder) we have been able to hire junior devs (in Colombia where rates are lower) and get the project going!

It is also a way to gauge interest in the project and provide a platform for funding future developments.

What do you guys think?

**Comment by xlotlu** [comment]
Well, sure, but right now the project is dying because of lack of steering. There's PRs going stale. There's issues and PRs that need discussing / a decision taken. Onboarding noobs (sorry, but that's the truth) is not gonna make a difference. Except maybe make matters worse by piling up more PRs that nobody's going to merge.

**Comment by goanpeca** [comment]
>Well, sure, but right now the project is dying because of lack of steering. 

Yes, this is concerning. Is there anything we can do about this? @nixjdm @lektor/developers ?

>Onboarding noobs (sorry, but that's the truth) is not gonna make a difference.

I think you are missing the point. Open collective is not going to fix any of the current issues on the short term, funds take some time to collect (it took over a full year before we could actually do something on the Spyder side). This is a medium/long term suggestion thinking about how t could help the project in the future.

Now how could we decide on new steering of the project, maybe adding more maintainers to the mix. I do not have much time for PR making, but I would be willing to help in project steering and management.

**Comment by runfalk** [comment]
I do agree with @xlotlu that the project is dying. There hasn't been a release since January 26th, 2019, and if I understand correctly that release wasn't based on current master, which means features that are even older (like #497) are still not part of a release. I have to admit that this has mellowed my interest in the project.

To get the project back on track there are a few urgent issues:

* There are currently multiple breaking bugs on master (#698, #720 and #635 which even has an unmerged fix).
* The CI system doesn't work properly because of constant changes to what pylint allows. #696 fixes this but is blocked by the 3.2 release. I don't think a 3.2 release should be a blocker for this.
* Make a new release with what we have on master. I have no idea which features that are part of 3.1.3 and not when browsing through the commit history. There is also unmerged documentation that's waiting for this (#279, #290).

In the slightly longer term I think we should release a lot more often, and to do that maybe more people should have PyPI access. Alternatively we should set up the CI to automatically release on new tags.

To adress the initial proposal of collecting funds. I don't think right now is the time to consider fundraising. Unless we can address the basic issues I outlined above there we won't do anything useful with funding.

**Comment by goanpeca** [comment]
I have tried to contact @nixjdm and I have not heard from him on any of the channels I tried.

We could try to do a "sprint" to update all things that are pending and assign among ourselves whatever needs reviewing before merging. Still if we do not hear from devs with admin access, we are kind of stranded here.

**Comment by runfalk** [comment]
I wouldn't worry too much about PyPI access. I assume @mitsuhiko still has access and can give that to someone if it comes down to that.

It seems like @yagebu has stepped up and fixed a sizable chunk of those issues. I'll try to review and merge their PRs today.

**Comment by yagebu** [comment]
I've been using Lektor for a few years and it's sad to see the project stagnate :( I haven't contributed any code so far but I've opened some PRs now and will try to add some more to get the master branch into a less broken state.

Edit: @runfalk: Thanks!

**Comment by nixjdm** [comment]
Sorry I haven’t responded to you sooner @goanpeca. I haven’t kept up with gitter in a while, but I did see your email. @xlotlu gave me a wellness check a while ago and I told him I’d get back soon, and then wasn’t able. I’ve been gone for a while, and as busy as I can handle in my life outside of Lektor for the last few months. I agree that lack of sufficient attention, by me mainly, is the biggest problem. I realized soon after I started helping with the project that its general health was closely tied with my activity, so I do feel responsible for the recent decline. The project revived a bit when I picked it up and was active, but consistency there is the key. I think it would revive more again if I/we became a lot more active, but it can’t really have a hope of becoming as popular as hugo / jekyll unless that amount of effort is a consistent minimum. I think I can give a little more attention for the next couple months, but it’s hard to see past the summer right now. I enjoy contributing, but I don’t have as much time for it as I used to.

Short term, you’re all right that open collective for junior devs won’t solve anything, but maybe long term it might help. It’s hard to say while we have our short term problems.

I’m totally willing to add pypi permissions and other maintainers if there’s anyone interested. I also mostly like releasing on tags. That has more pros than risks probably, so I’m fine with that too.

#698 should not make it into a release - which was just fixed (Thanks @yagebu!). I should have released something prior to that, but that’s past. #720 can be changed outside a release since that code is pulled from getlektor. Trying to reread #635, I think that’s in the same situation. Another issue with our CI is that those install scripts aren’t tested anywhere. There’s certainly a lot of tech debt, but with #698 now out of the way we can probably cut a release.

**Comment by nixjdm** [comment]
Sorry I haven’t responded to you sooner @goanpeca. I haven’t kept up with gitter in a while, but I did see your email. @xlotlu gave me a wellness check a while ago and I told him I’d get back soon, and then wasn’t able. I’ve been MIA for a while, and as busy as I can handle in my life outside of Lektor for the last few months. I agree that lack of sufficient attention, by me mainly, is the biggest problem. I realized soon after I started helping with the project that its general health was closely tied with my activity, so I do feel responsible for the recent decline. The project revived a bit when I picked it up and was active, but consistency there is the key. I think it would revive more again if I/we became a lot more active, but it can’t really have a hope of becoming as popular as hugo / jekyll unless that amount of effort is a consistent minimum. I think I can give a little more attention for the next couple months, but it’s hard to see past the summer right now. I enjoy contributing, but I don’t have as much time for it as I used to.

Short term, you’re all right that open collective for junior devs won’t solve anything, but maybe long term it might help. It’s hard to say while we have our short term problems.

I’m totally willing to add pypi permissions and other maintainers if there’s anyone interested. I also mostly like releasing on tags. That has more pros than risks probably, so I’m fine with that too.

#698 should not make it into a release - which was just fixed (Thanks @yagebu!). I should have released something prior to that, but that’s past. #720 can be changed outside a release since that code is pulled from getlektor. Trying to reread #635, I think that’s in the same situation. Another issue with our CI is that those install scripts aren’t tested anywhere. There’s certainly a lot of tech debt, but with #698 now out of the way we can probably cut a release.

**Comment by goanpeca** [comment]
Thanks for your words @nixjdm I can help with release process and updating CI as needed.

I will take the weekend to go over all the open issues and add some comments to see what could be a priority in order to try to organize the work a bit. I am a user of this project and I really like it so I want to see it thrive :-).

@lektor/developers would you like to set up a meeting (30 mins?) sometime next week to define some minimal roadmap going forward and maybe distribute a bit tasks?

I have some bandwidth at the moment that I can devote to help with the project management.

**Comment by roberto-arista** [comment]
hey there!
heavy user here

I rely on Lektor for many projects, and it 99% of the time my choice when I need to publish something on the web. I don't think I can be helpful immediately, but I am up for funding or helping with reviving the project website (or other design-related issues)

👋

**Comment by bbbart** [comment]
Hi all,

I also want to let the community know about another person heavily relying on Lektor for online publishing.

I also happen to be a technical guy with some Python experience behind him. Of course, a lack of time is also keeping me from promising too much direct involvement soon, but perhaps I can prove to be helpful one day as well.

Thank you for the great work and loads of time and effort this project has saved me from in my online endeavours.

**Comment by xlotlu** [comment]
Do we consider this issue closed? Is the project future settled and safe enough? :) 

Maybe we should create a ROADMAP file to both structure / prioritise our plans, and be transparent about them.

**Comment by goanpeca** [comment]
>Maybe we should create a ROADMAP file to both structure / prioritise our plans, and be transparent about them.

We need a Vision, then a Roadmap :-)

**Comment by dairiki** [comment]
A year on...

What is the current state and mindset of project maintenance/maintainers?

It's been a year and over 300 commits since the last release was made.
There are basic, simple PRs lingering open on `lektor-website`, and it seems like it has been some time since the getlektor.com website has seen any updates.

I don't mean to address blame at any particular party — we are all volunteers — but it's not a great state of affairs for a very useful project that seems still to be actively used by quite a number of people.

It seems likely that right now we don't have the manpower or interest to accomplish many major changes — things like major new features, or combining the docs with the main repo, etc. — in an organized manner.  Still, bugs are being fixed, and the master branch has seen quite a bit of love since the last release.  It would be nice (and healthy for the project, in terms of encouraging new users) to have more frequent releases.  (Why not release after each bug fix?)

What's the current collective state of mind on this?

Who can make it happen?  (Who has the keys to PyPI and getlektor.com?)

(I'm willing to help, as long as it doesn't require a major time commitment.)

**Comment by skorokithakis** [comment]
As far as releases go, I noticed I made many more releases in my personal projects when I automated the process. It's amazing how much automation takes the hassle out of releasing, now I just tag master with the version and everything else is handled automatically by CI.

Maybe that would be useful for this project as well.

**Comment by dairiki** [comment]
> As far as releases go, I noticed I made many more releases in my personal projects when I automated the process.

Yes, I've worked on projects where this is done and it works well.
It is a little scary to remove too many safety checks from the release process since it is hard to undo a botched PyPI release (especially, say, a botched version number.)

**Comment by skorokithakis** [comment]
Hmm, what sort of safety checks are you thinking? You can make release tags protected so that only certain people can release.

**Comment by dairiki** [comment]
> Hmm, what sort of safety checks are you thinking? You can make release tags protected so that only certain people can release.

I just meant that when releasing manually, there are more steps and so (perhaps) more chance to catch mistakes before publishing to PyPI.  A couple of things that make me nervous about automating the release process:
- Fat fingering a version number (e.g. release a 5.4 rather when one intended to release 4.5)
- Having the CI script break in a way that produces a broken release.

I'm probably overly paranoid.  I suppose that one can also argue the reverse just as easily: that having an automated release process makes botched release less likely.  (E.g. it makes it "impossible" to forget to build the JS for the GUI before making the release.)

**Comment by skorokithakis** [comment]
Hmm, I agree that there are more steps to catch mistakes, but the mistakes are usually caused by the steps themselves. With the way I do it, you tag the release number, check it, and if it's okay, you push. There isn't much that can go wrong there.

**Comment by dairiki** [comment]
As long as releases get made, I will be happy!

I think we're getting a bit off track.  My main queries were:

Who is able to make releases?  (And publish to getlektor.com?)
Is there maybe have a reason for not having made a release in a year?  Are we waiting for something?  Or have those with the keys just lost interest in the project?

**Comment by andoresuperesu** [comment]
I've previously seen @runfalk and @nixjdm work on releases. Obviously @mitsuhiko himself could. There also used to be meetings every month at least. However those stopped on July past year. I'd hope for some participation from them. @goanpeca is a prominent contributor to other libraries, his input in project management as suggested here might be extremely useful. I can volunteer some writing if somewhat guided on the technical side.

**Comment by nixjdm** [comment]
About releases, note you can see who has the ability to release on the [pypi page](https://pypi.org/project/Lektor/), and in particular the "getlektor" user, which should be a pypi user on each of the organization repos that are published packages. Some work was done to make releasing easier in [lektor-atom](https://github.com/lektor/lektor-atom/blob/master/.github/workflows/publish.yml), that would use pypi keys associated with the "getlektor" user to publish via tagging. The intention was that once we were sufficiently happy with that approach (and the code to do it), we could copy & paste the code into this and all of the other repos. This repo currently has keys loaded do to that as well, ready for the github actions code to use them. With this in place anyone who can push a tag could create a release.

@skorokithakis I'm not sure how your CI might differ from this. @goanpeca was mostly the one leading this effort. I can't recall what if anything was left that we needed before migrating the lektor-atom release code over here. I _thought_ there was something not quite right yet, something about tests interfering with releasing? I'm not sure. Either way, I support release-by-tagging.

In lieu of release-by-tagging being done in this repo, I can create a release. Or I can give that permission to another maintainer who doesn't have it yet. @yagebu or @skorokithakis maybe? If I do it, I'd like to ask the community for help at least in updating the changelog before I release. If we need, we can make a new issue to discuss that. I'm also open to adding new maintainers if anyone would like to help. Please email me if you're interested.

Now for the main point of this issue thread, I can say that I have not been active lately. Over the past few years I've had increasingly less time for Lektor, and that's unlikely to substantially change soon. So I don't feel I should be terribly opinionated right now if I can't be more hands-on. I mainly want to enable others, and participate as much as I can as long as it doesn't get in the way.

**Comment by dairiki** [comment]
> I'd like to ask the community for help at least in updating the changelog before I release.

I've started work on this.  I will keep working on it over the next few days...  See #929.

**Comment by goanpeca** [comment]
Hello :) , sorry, life happened and I lost track of Lektor. I am still using for some websites (but I have not updated them in a while!)

**Comment by yagebu** [comment]
Thank you @dairiki for getting this issue going again and pushing for a release (and all your other recent contributions as well). I think between the two of us, we've done enough occasional work on Lektor in the last year to bring in the odd bugfix and improvement to keep the dev version in a working state. I agree that we're really lacking a release right now, since the last release contains some bugs and is a pain to install due to a couple of breaking changes in dependencies. A couple of months ago, I already prepared a 3.2.1 bugfix release but none of the maintainers with the appropriate permissions had the time back then to do it I believe. I believe long-term automatic releases from CI could be nice but I don't have a ton of experience with them. I've usually had a make target that does most of the work, so I'll try to implement that first.

@nixjdm: I'd be happy to make releases :)

**Comment by paternal** [comment]
It's great to see that there are a few people who can keep this project alive! 🎉 It would be great to publish a new release to https://github.com/lektor/lektor-atom too. For instance, a dependency problem has been fixed, but not released yet (lektor/lektor-atom#32).
Thanks!

**Comment by dairiki** [comment]
> Please email me if you're interested.

@nixjdm I sent you an email last week.  I have not heard back.  Did you receive it?

**Comment by dairiki** [comment]
What more is it going to take to get a release out the door?

A (totally reasonable) request for the changelog to be updated before making a release.  I spent an afternoon and did that. 

Now work seems to be proceeding mostly on various refactors and code reformatting projects.  Things like the admin JS refactor are great, but are likely to introduce new buglets.  Let's get a working release out the door before we merge these things.

Many of us have been running lektor from git master for a while now, since 3.2.0, over a year old now, is broken.  As such, it is fairly well tested in use, and works, at least, for some of us.  Before we make too many changes to current master, let's make a release!

It seems crazy that, on one hand, we want this to be a user-friendly tool that our parents can use, while at the same time saying "Yeah, that's a known bug.  It was fixed in master (a year ago).  You'll have to run the git version of the code to get around that."

**Comment by yagebu** [comment]
> What more is it going to take to get a release out the door?

3.2.2 is out already ;) (https://pypi.org/project/Lektor/)

But I agree that we should also soon have a release of the master branch soon. However there's some breaking changes suggested for 4.0 that would still take some time to get done, so I'm planning to get a 3.3 release based on the master branch out soon. The only thing I'd still like to include is #933, could you get that ready @dairiki?

**Comment by dairiki** [comment]
I will work on #933 on Monday, if not before.

I would guess that dropping support for Python < 3.5 is "breaking" enough that the next release (after 3.2.x) should be 4.0 (assuming that we are committed to SemVer.)
 
What other breaking changes are being considered?

**Comment by nixjdm** [comment]
@dairiki Sorry, I missed it, but thanks for the extra ping. You're in!

**Comment by yagebu** [comment]
> I would guess that dropping support for Python < 3.5 is "breaking" enough that the next release (after 3.2.x) should be 4.0 (assuming that we are committed to SemVer.)

IMHO dropping Python 2 (or early Python 3 users) is not a breaking change according to semver and thanks to `python_requires` (at least how I understand it should work) it shouldn't really be breaking in any other sense to Python 2 users, right? They'll just never "see" that package being installed by their package manager.

> What other breaking changes are being considered?

Mainly the deprecation cleanups that I've added to the corresponding milestone. And I'd really want the website to work again before we announce a major release (I've asked around for getting deployments to work again, but getting those accesses worked out might just take some time).

**Comment by dairiki** [comment]
> You're in!

@nixjdm  Thank you!

**Comment by andoresuperesu** [comment]
@yagebu would it be counterproductive to have a secondary domain with the updated page in the mean time? I'd be pleased to provide it if deemed appropriate. It wouldn't be fair to let something outside the projects scope halt it.

**Comment by dairiki** [comment]
@yagebu I guess the deploy secrets referenced in [deploy.yml](https://github.com/lektor/lektor-website/blob/42f74d7e3e896dda6e5d11c21ddf7e80c125d0a7/.github/workflows/deploy.yml#L31) don't work (or don't exist)?

**Comment by paternal** [comment]
> @yagebu would it be counterproductive to have a secondary domain with the updated page in the mean time? I'd be pleased to provide it if deemed appropriate. It wouldn't be fair to let something outside the projects scope halt it.

If the secondary domain is only temporary, wouldn't the github pages default domain https://lektor.github.io/lektor-website be sufficient?

**Comment by andoresuperesu** [comment]
> If the secondary domain is only temporary, wouldn't the github pages default domain https://lektor.github.io/lektor-website be sufficient?

Technically yes, but temporary might be counted in years. I think I read about that issues with the web page on 2019 though.

**Comment by mitsuhiko** [comment]
If someone needs a subdomain set up please let me know and I can create a record to point it somewhere.

**Comment by dairiki** [comment]
> If someone needs a subdomain set up please let me know and I can create a record to point it somewhere.

Thank you.

I'm confused as to who has the keys to what.  Right now lektor-website is published at www.getlektor.com which is served by flow.srv.pocoo.org.  I had assumed that all of that (both the getlektor.com domain and the pocoo.org server) was under the control of @mitsuhiko, but maybe not?

The [deploy.yml](https://github.com/lektor/lektor-website/blob/42f74d7e3e896dda6e5d11c21ddf7e80c125d0a7/.github/workflows/deploy.yml#L31) for lektor-website implies that there were (at least at one time, or in planning) secrets configured for the repo which allowed publishing updates to the website.  If those don't currently work, can we fix them?

Maybe we should just point www.getlektor.com to the github pages for lektor-website?  (That, still, I think, would take help from whoever controls the getlektor.com domain, and an admin of the lektor-website repo.)

**Comment by yagebu** [comment]
> The deploy.yml for lektor-website implies that there were (at least at one time, or in planning) secrets configured for the repo which allowed publishing updates to the website. If those don't currently work, can we fix them?

I think those secrets are for the (attempted) GH pages deployment. I believe the working secrets for getlektor.com were only on the now defunct travis-ci.org. It seems to me that for getting deployments to flow.srv.pocoo.org and hence getlektor.com working again, we only need to have those secrets added to the GH repo.

**Comment by dairiki** [comment]
@yagebu Bump. I would still like to see a release of the master branch (whether that's 3.3 or 4.0).  What is that currently blocking on?

**Comment by yagebu** [comment]
Thanks for the bump @dairiki - I had wanted to wait for a working website and then Lektor fell off my radar a bit. There's no blocker IMHO. I'll try to get 3.3 out in the next few days :)

**Comment by dairiki** [comment]
@yagebu Awesome, thank you!

I'll try to make sure that CHANGES is up-to-date on master over the next couple of days.

Can we maybe set up a publish-to-pypi workflow so that any maintainer can create a release?  I believe this was discussed before, but I can't find the reference atm.  I can try to create the workflow, if you'd like.  (But someone else will have to create/set the secret access token.)

**Comment by yagebu** [comment]
I've set up an automatic publishing CI flow on tag (https://github.com/lektor/lektor/blob/master/.github/workflows/publish.yml) and used it to release 3.3.0 :)
